regex/src/main/scala/regex/Regex.scala

```diff
@@ -15,13 +24,46 @@ package regex
   */
 
 /** Simplifies a regular language */
-def simplify(lang: RegularLanguage): RegularLanguage = ???
+def simplify(lang: RegularLanguage): RegularLanguage = lang match
+  // base case
+  case (Star(Epsilon)) => Epsilon
+  case (Star(Empty)) => Empty
+
+  // recursive cases
+  case Concat(Empty, _) | Concat(_, Empty) => simplify(Empty)
+
+  case Concat(Epsilon, lang) => simplify(lang)
+  case Concat(lang, Epsilon) => simplify(lang)
+  case Union(Empty, lang) => simplify(lang)
+  case Union(lang, Empty) => simplify(lang)
+
+  case Concat(firstLang, secondLang) => Concat(simplify(firstLang), simplify(secondLang))
+  case Union(firstLang, secondLang) => Union(simplify(firstLang), simplify(secondLang))
+  case Star(lang) => Star(simplify(lang))
+
+  case _ => lang
 
 /** A language is nullable if it contains ε */
-def nullable(lang: RegularLanguage): Boolean = ???
+def nullable(lang: RegularLanguage): Boolean = lang match
+  case Empty | Character(_) => false
+  case Epsilon | Star(_) => true
+
+  case Concat(firstLang, secondLang) => nullable(firstLang) & nullable(secondLang)
+  case Union(firstLang, secondLang) => nullable(firstLang) | nullable(secondLang)
+  case _ => nullable(simplify(lang))
+
+
 
 /** Computes the derivative of a language, with respect to a character */
-def derivative(l: RegularLanguage)(c: Char): RegularLanguage = ???
+def derivative(l: RegularLanguage)(c: Char): RegularLanguage = l match
+  case Character(d) if d == c => Epsilon
+  case Empty | Epsilon | Character(_) => Empty
+
+  case Union(firstLang, secondLang) => Union(derivative(firstLang)(c), derivative(secondLang)(c))
+  case Concat(firstLang, secondLang) if !nullable(firstLang) => Concat(derivative(firstLang)(c), secondLang)
+  case Concat(firstLang, secondLang) => Union(Concat(derivative(firstLang)(c), secondLang), derivative(secondLang)(c))
+  case Star(lang) => Concat(derivative(lang)(c), Star(lang))
```

Comment:
Like in the first part, your code is very readable and easy to understand. You also made good use of the underscores. I had never done pattern matching before, but I like the way Scala handles it.